lucet-wasi/src/hostcalls.rs

```diff
@@ -980,6 +988,241 @@ pub extern "C" fn __wasi_random_get(
     return wasm32::__WASI_ESUCCESS;
 }
 
+// define the `fionread()` function, equivalent to `ioctl(fd, FIONREAD, *bytes)`
+nix::ioctl_read_bad!(fionread, nix::libc::FIONREAD, c_int);
+
+fn wasi_clock_to_relative_ns_delay(
+    wasi_clock: host::__wasi_subscription_t___wasi_subscription_u___wasi_subscription_u_clock_t,
+) -> u128 {
+    if wasi_clock.flags != wasm32::__WASI_SUBSCRIPTION_CLOCK_ABSTIME {
+        return wasi_clock.timeout as u128;
+    }
+    let now: u128 = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("Current date is before the epoch")
+        .as_nanos();
+    let deadline = wasi_clock.timeout as u128;
+    deadline.saturating_sub(now)
+}
+
+#[derive(Debug, Copy, Clone)]
+struct ClockEventData {
+    delay: u128,
+    userdata: host::__wasi_userdata_t,
+}
+#[derive(Debug, Copy, Clone)]
+struct FdEventData {
+    fd: c_int,
+    type_: host::__wasi_eventtype_t,
+    userdata: host::__wasi_userdata_t,
+}
+
+fn __wasi_poll_oneoff_handle_timeout_event(
+    vmctx: &mut Vmctx,
+    output_slice: &mut [wasm32::__wasi_event_t],
+    nevents: wasm32::uintptr_t,
+    timeout: Option<ClockEventData>,
+) -> wasm32::__wasi_errno_t {
+    if let Some(ClockEventData { userdata, .. }) = timeout {
+        let output_event = host::__wasi_event_t {
+            userdata,
+            type_: wasm32::__WASI_EVENTTYPE_CLOCK,
+            error: wasm32::__WASI_ESUCCESS,
+            u: host::__wasi_event_t___wasi_event_u {
+                fd_readwrite: host::__wasi_event_t___wasi_event_u___wasi_event_u_fd_readwrite_t {
+                    nbytes: 0,
+                    flags: 0,
+                },
+            },
+        };
+        output_slice[0] = enc_event(output_event);
+        if let Err(e) = unsafe { enc_pointee(vmctx, nevents, 1) } {
+            return e;
```

Comment:
needs to be `enc_errno(e)` whenever we return an error from an `enc_` or `dec_`. it would be nice if bindgen could make these as newtype structs, so we could have the compiler remind us to do this :frowning_face:

Reply:
done